File: app/src/main/java/com/example/shibeapplication/views/MainActivity.java
package com.example.shibeapplication.views;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.shibeapplication.R;
import com.example.shibeapplication.adapters.ShibeAdapter;
import com.example.shibeapplication.viewmodels.ShibeViewModel;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    @BindView(R.id.rv_shibeContainer) RecyclerView rvShibes;
    @BindView(R.id.btn_switch) Button btnSwitch;
    private ShibeViewModel viewModel;
    private GridLayoutManager manager;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //view and viewModel instantiation
        ButterKnife.bind(this);
        viewModel = new ShibeViewModel();

        //set Event Listeners
        viewModel
                .getShibes(20)
                .enqueue(new Callback<List<String>>() {
                    @Override
                    public void onResponse(Call<List<String>> call, Response<List<String>> response) {

                        ShibeAdapter adapter = new ShibeAdapter(response.body());
                        rvShibes.setAdapter(adapter);
                    }

                    @Override
                    public void onFailure(Call<List<String>> call, Throwable t) {
                        Toast.makeText(MainActivity.this, "Image Fetch failed.", Toast.LENGTH_SHORT).show();
                        Log.e("FailedImageFetch","Failed to retrieve images of Shibes");
                        t.printStackTrace();
                    }
                });
        btnSwitch.setOnClickListener(this);

        //Set the layout manager
        manager = new GridLayoutManager(this, 1);
        rvShibes.setLayoutManager(manager);


    }

    @Override
    public void onClick(View view) {
        int spanCount = manager.getSpanCount();
        if (spanCount == 1){
            manager.setSpanCount(4);
        }else if(spanCount == 4){
            manager.setSpanCount(1);
        }
    }
}

File: app/src/main/java/com/example/shibeapplication/views/MainActivity.kt
package com.example.shibeapplication.views

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.shibeapplication.databinding.ActivityMainBinding
import com.example.shibeapplication.viewmodels.ShibeViewModel
import com.example.shibeapplication.views.adapters.ShibeAdapter
import dagger.hilt.android.AndroidEntryPoint
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject

@AndroidEntryPoint
class MainActivity : AppCompatActivity(), View.OnClickListener {

    @Inject
    lateinit var viewModel: ShibeViewModel

    private var gridManager: GridLayoutManager? = null
    private var linearManager: LinearLayoutManager? = null

    @Inject
    lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        //view and viewModel instantiation

        //set Event Listeners
        viewModel
                .getShibes(20)
                .enqueue(object : Callback<List<String>> {
                    override fun onResponse(call: Call<List<String>>, response: Response<List<String>>) {
                        response.body()?.let {
                            binding.rvShibeContainer.adapter = ShibeAdapter(it)
                        }
                    }

                    override fun onFailure(call: Call<List<String>>, t: Throwable) {
                        Toast.makeText(this@MainActivity, "Image Fetch failed.", Toast.LENGTH_SHORT).show()
                        Log.e("FailedImageFetch", "Failed to retrieve images of Shibes")
                        t.printStackTrace()
                    }
                })
        binding.btnSwitch.setOnClickListener(this)

        //Set the layout manager
        gridManager = GridLayoutManager(this, 3)
        linearManager = LinearLayoutManager(this)

        binding.rvShibeContainer.layoutManager = gridManager
    }

    override fun onClick(view: View) {
        with(binding.rvShibeContainer) {
            when (layoutManager) {
                is GridLayoutManager -> {
                    layoutManager = linearManager
                }
                is LinearLayoutManager -> {
                    layoutManager = gridManager
                }
            }
        }
    }
}